presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java

```diff
@@ -133,32 +192,19 @@ public Block readStruct(Type type, List<String> path)
             NamedTypeSignature namedTypeSignature = parameters.get(i).getNamedTypeSignature();
             Type fieldType = typeManager.getType(namedTypeSignature.getTypeSignature());
             String name = namedTypeSignature.getName();
-            path.add(name);
-            Optional<RichColumnDescriptor> columnDescriptor = getDescriptor(fileSchema, requestedSchema, path);
-            if (!columnDescriptor.isPresent()) {
-                path.remove(name);
-                blocks[i] = RunLengthEncodedBlock.create(type, null, batchSize);
-                continue;
-            }
-
-            if (ROW.equals(fieldType.getTypeSignature().getBase())) {
-                blocks[i] = readStruct(fieldType, path);
-            }
-            else {
-                blocks[i] = readPrimitive(columnDescriptor.get(), fieldType);
-            }
-            path.remove(name);
+            blocks[i] = readBlock(name, fieldType, path, new ArrayList<>(), positions);
         }
 
         InterleavedBlock interleavedBlock = new InterleavedBlock(blocks);
-        int[] offsets = new int[batchSize + 1];
+        int blockSize = blocks[0].getPositionCount();
+        int[] offsets = new int[blockSize + 1];
         for (int i = 1; i < offsets.length; i++) {
             offsets[i] = i * parameters.size();
         }
-        return new ArrayBlock(batchSize, new boolean[batchSize], offsets, interleavedBlock);
+        return new ArrayBlock(blockSize, new boolean[blockSize], offsets, interleavedBlock);
     }
 
-    public Block readPrimitive(ColumnDescriptor columnDescriptor, Type type)
+    public Block readPrimitive(ColumnDescriptor columnDescriptor, Type type, List<Integer> offsets, List<Integer> positions)
```

Comment:
There is something strange with this method name.  `ParquetBlockLoader` calls this for all reads, so I assume this works for all types.

Additionally, I would make this `protected` and add a `public` variant that does not take the list arguments